gctree/dag_filtering.py

```diff
@@ -0,0 +1,507 @@
+        plt.tick_params(
+            axis="x", which="both", bottom="off", top="off", labelbottom="off"
+        )
+        plt.savefig(outbase + ".inference.likelihood_rank." + img_type)
+
+    if verbose:
+        print(f"params: {(p, q)}")
+        print("stats for optimal trees:")
+        print(
+            "alleles\tlogLikelihood\t\tisotype_parsimony\tmutability_parsimony"
+            + ("\ttreescore" if priority_weights else "")
+        )
+        # with format (ll, _, isotypepars, _, mutabilitypars, alleles)
+        stats = trimdag.optimal_weight_annotate(
+            **dagweight_kwargs, optimal_func=lambda l: min(l, key=minfunckey)
+        )
+        print(
+            f"{stats[5]}\t{stats[0]}\t{stats[2]}\t\t\t{stats[4]}\t"
+            + str(minfunckey(stats) if priority_weights else "")
+        )
+
+    return trimdag
+
+
+def ll_cmcount_dagfuncs(p: np.float64, q: np.float64):
+    """A slower but more numerically stable equivalent to ll_genotype_dagfuncs.
+
+    Args:
+        p, q: branching process parameters
+
+    Returns:
+        A ``historydag.utils.AddFuncDict`` object, which can be unpacked and used
+        as keyword arguments for DAG trimming and weight annotation according to
+        branching process likelihood.
+        Weight format is ``(log likelihood, cmcounts)`` where cmcounts is a FrozenMultiset containing abundance, mutant clade count pairs.
+        To use these functions for DAG trimming, use an optimal function like
+        `lambda l: max(l, key=lambda ll: ll[0])` for clarity, although min or max should work too.
+    """
+
+    funcdict = bp._cmcounter_dagfuncs()
+    cmcount_weight_func = funcdict["edge_weight_func"]
+    cmcount_addfunc = funcdict["accum_func"]
+
+    @lru_cache(maxsize=None)
+    def ll(cmcounter: FrozenMultiset) -> float:
+        if cmcounter:
+            if (0, 1) in cmcounter:
+                cmcounter = cmcounter - {(0, 1)} + {(1, 1)}
+            return bp._lltree(tuple(cmcounter.items()), p, q)[0]
+        else:
+            return 0.0
+
+    def edge_weight_func(n1, n2):
+        st = cmcount_weight_func(n1, n2)
+        return (ll(st), st)
+
+    def accum_func(cmsetlist: List[Tuple[float, FrozenMultiset]]):
+        st = cmcount_addfunc([t[1] for t in cmsetlist])
+        return (ll(st), st)
+
+    return hdag.utils.AddFuncDict(
+        {
+            "start_func": lambda n: (0, FrozenMultiset()),
+            "edge_weight_func": edge_weight_func,
+            "accum_func": accum_func,
+        },
+        names=("LogLikelihood", "cmcounters"),
+    )
+
+
+def ll_genotype_dagfuncs(p: np.float64, q: np.float64):
+    """Return functions for counting tree log likelihood on the history DAG.
+
+    Although these functions are fast, for numerical consistency use
+    :meth:`ll_cmcount_dagfuncs` instead.
+
+    Args:
+        p, q: branching process parameters
+
+    Returns:
+        A ``historydag.utils.AddFuncDict`` object, which can be unpacked and used
```

Comment:
Can we cross-reference to hdag docs here?

Reply:
Yes. Those docs aren't up on GitHub pages yet, but I'll fix that.